src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal class SwitchExpressionBinder : Binder
+    {
+        protected readonly SwitchExpressionSyntax SwitchExpressionSyntax;
+
+        private BoundExpression _switchGoverningExpression;
+        private DiagnosticBag _switchGoverningDiagnostics;
+
+        internal SwitchExpressionBinder(SwitchExpressionSyntax switchExpressionSyntax, Binder next)
+            : base(next)
+        {
+            SwitchExpressionSyntax = switchExpressionSyntax;
+        }
+
+        internal override BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax node, Binder originalBinder, DiagnosticBag diagnostics)
+        {
+            // Bind switch expression and set the switch governing type.
+            var boundSwitchGoverningExpression = SwitchGoverningExpression;
+            diagnostics.AddRange(SwitchGoverningDiagnostics);
+            ImmutableArray<BoundSwitchExpressionCase> switchCases = BindSwitchExpressionCases(node, originalBinder, diagnostics);
+            bool hasErrors = false;
+            TypeSymbol resultType = InferResultType(switchCases, diagnostics);
+            switchCases = AddConversionsToCases(switchCases, resultType, diagnostics);
+            // PROTOTYPE(patterns2): check for subsumption, completeness, etc.
+            //var hasErrors = CheckSwitchErrors(node, boundSwitchGoverningExpression, switchCases, decisionDag, diagnostics);
+            return new BoundSwitchExpression(node, boundSwitchGoverningExpression, switchCases, resultType, hasErrors);
+        }
+
+        /// <summary>
+        /// Infer the result type of the switch expression by looking for a common type.
+        /// </summary>
+        private TypeSymbol InferResultType(ImmutableArray<BoundSwitchExpressionCase> switchCases, DiagnosticBag diagnostics)
+        {
+            var seenTypes = new HashSet<TypeSymbol>();
+            var typesInOrder = ArrayBuilder<TypeSymbol>.GetInstance();
+            foreach (var @case in switchCases)
+            {
+                var type = @case.Value.Type;
+                if (type != null && seenTypes.Add(type))
+                {
+                    typesInOrder.Add(type);
+                }
+            }
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var commonType = BestTypeInferrer.InferBestType(typesInOrder.ToImmutableAndFree(), Conversions, ref useSiteDiagnostics);
```

Comment:
The problem will be less for this construct, because the type can come from any one of the arms of the switch expression.

---
In reply to: [166481222](https://github.com/dotnet/roslyn/pull/23357#discussion_r166481222) [](ancestors = 166481222)